libsolidity/ast/TypeProvider.cpp

```diff
@@ -0,0 +1,395 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/ast/TypeProvider.h>
+#include <libdevcore/ArrayUtil.h>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+using namespace std;
+using namespace dev;
+using namespace solidity;
+
+template <size_t... N>
+constexpr std::array<IntegerType, sizeof...(N)> createIntegerTypes(IntegerType::Modifier _modifier, std::index_sequence<N...>)
+{
+	return makeArray<IntegerType>(IntegerType((static_cast<unsigned>(N) + 1) * 8, _modifier)...);
+}
+
+template <size_t... N>
+constexpr std::array<FixedBytesType, sizeof...(N)> createFixedBytesTypes(std::index_sequence<N...>)
+{
+	return makeArray<FixedBytesType>(FixedBytesType(static_cast<unsigned>(N) + 1)...);
+}
+
+BoolType const TypeProvider::m_boolType{};
```

Comment:
Is this required to be mentioned here?

Reply:
static class variables must be instanciated somewhere, as they cannot be initialized in the class declaration.

Reply:
actually, just checked. if BoolType() (etc) would be `constexpr`, then I could have put it into the header and skip the explicit instanciation outside the class decl here.